drivers/bluetooth/hci/userchan.c

```diff
@@ -93,6 +95,61 @@ static struct net_buf *get_rx(const uint8_t *buf)
 	return NULL;
 }
 
+/**
+ * @brief Decode the length of an HCI H4 packet
+ * @details Decodes packet length according to Bluetooth spec v5.2 Vol 4 Part E
+ * @param packet_type	Type of packet according to Bluetooth spec v5.2 Vol 4, Part A Table 2.1
+ * @param packet_buff	Pointer to packet buffer
+ */
+static uint16_t decode_hci_packet_length(uint8_t packet_type, const uint8_t *packet_buff)
+{
+	uint16_t payload_size = 0;
+	uint8_t header_size = 0;
+	const uint8_t packet_type_len = sizeof(packet_type);
+
+	switch (packet_type) {
+	case H4_CMD:
+		const struct bt_hci_cmd_hdr *hdr = &packet_buff[sizeof(packet_type)];
+
+		/* Parameter Total Length */
+		payload_size = hdr->param_len;
+		header_size = BT_HCI_CMD_HDR_SIZE;
+		break;
+	case H4_ACL:
+		const struct bt_hci_acl_hdr *hdr = &packet_buff[sizeof(packet_type)];
+
+		/* Data Total Length */
+		payload_size = sys_get_le16(&hdr->len);
+		header_size = BT_HCI_ACL_HDR_SIZE;
+		break;
+	case H4_SCO:
+		const struct bt_hci_sco_hdr *hdr = &packet_buff[sizeof(packet_type)];
+
+		/* Data_Total_Length */
+		payload_size = hdr->len;
+		header_size = BT_HCI_SCO_HDR_SIZE;
+		break;
+	case H4_EVT:
+		const struct bt_hci_evt_hdr *hdr = &packet_buff;
```

Comment:
This doesn't seem correct. You have the packet type as the first byte and the packet-specific header comes after that. In all other cases you read `&packet_buff[1]` but here you read `&packet_buff[0]`. Actually, I don't see any reason why you need two parameters to this function. Just pass the original buffer and inside the function you can then do `uint8_t type = buf[0];`.

Reply:
Yes, sorry mistake on my part while changing to the suggestions of the other reviewer.